Various changes for tools/check-typo

dra27 opened the pull request:
This is a prerequisite for #1148 moving towards having the entire repository pass the `tools/check-typo` script.

Review points definitely required:
 - @damiendoligez - I've altered `tools/check-typo` to permit UTF-8 characters in copyright and authors lines in the headers (allows several existing headers to retain accented characters and also means no author in future should feel compelled to use incorrect/alternate spellings of names just to satisfy a script)
 - I've used a URL shortener to reduce some line-lengths in the documentation, which may be contentious for several reasons (@gasche: thoughts?)
 - The two changes in `Makefile` are worthwhile in their own right, but consequently should be reviewed as code, not reformatting
 - I'm reasonably confident that `CCOUTPUT` in `otherlibs/systhreads/Makefile` is a merge artefact from @shindere's marvellous merge, but I didn't triple-check it

There are various people who should briefly sign off on (hopefully) trivial alterations:
 - [x] @stedolan - `tools/check-symbol-names` was lacking a copyright and I assume that you're happy to be credited with your script :wink:
 - [x] @edwintorok - `tools/lintapidiff.ml` has a mis-formatted header, but I believe you are the both author and copyright holder!
 - [x] @nojb - `tools/make_opcodes.mll` has a wrong licence, either because it was miscopied or possibly because the patch originally predated the change of OCaml's licence. Please confirm that switching the file to declare LGPL 2.1 as for the rest of the tree is OK.

dra27 committed:
Missing testsuite entries in .gitignore

dra27 committed:
Tweak .gitattributes for fakeclock.c in testsuite

Large piece of code which does have a header.

dra27 committed:
Permit UTF-8 characters in copyright and authors

Allows names to include accented characters, but only in a comment.

dra27 committed:
Correct licence for tools/make_opcodes.mll

Incorrectly copied from an old QPL header.

dra27 committed:
Correct copyright header

dra27 committed:
Add missing copyright header

dra27 committed:
Reformat MD5 test

Reduce line lengths and tidy-up the code while at it.

dra27 committed:
Remove unused CCOUTPUT in systhreads Makefile

dra27 committed:
Refactor byterun/Makefile for long-line

dra27 committed:
Correct copyright header formatting

dra27 committed:
Tidy-up generation of utils/config.ml

Simultaneously deals with an overly-long line and also ensure that
backslashes would be escaped no matter which variables they happened to
appear in.

dra27 committed:
Refactor lintapidiff command for long-line

dra27 committed:
Shorten URLs as much possible.

dra27 committed:
Remove non-ASCII characters

dra27 committed:
Refactor configuration steps in AppVeyor script

nojb commented:
@dra27 yes, it was miscopied, change is OK.

gasche commented:
Indeed, I don't like the URL shortener idea -- thanks for asking. We win very little (the only strong positive I see is to respect the 80 columns limit), and we start depending on an extra external service that (1) introduces a new way our links may become dead (although it's reasonable to hope that even if Google killed the service (likely), it would preserve existing URLs) and (2) leaks information from people reading documentation to a third-party for no good reason¹.

One thing that would be useful, I think, is to ask archive.org for a version of these pages, which has the side-effect of getting it to record them. (This is not failsafe, if the site later uses a robots.txt to disable archiving they will still be unavailable.)

¹: for the record, although I like privacy and believe that google is probably evil or may be in the future, I don't personally mind that much the use of a Google-supported (or someone-else-supported) url shortener, but I know that some people do mind; in particular the German-speaking school of free software and privacy activism is very sensitive about privacy issues, even little things, and I think it's nice to avoid moves that could turn them off OCaml contributions.

Octachron commented on testsuite/tests/lib-bigarray/change_layout.ml:
```diff
@@ -4,7 +4,7 @@ open Bigarray
 let pp_sep ppf () = Format.fprintf ppf ";@ "
  let print_array pp ppf a =
- Format.fprintf ppf "@[<hov>⟦%a⟧@]"
+ Format.fprintf ppf "@[<hov>\xe2\x9f\xa6%a\xe2\x9f\xa7@]"
```
May I suggest to replace the format string with a simple "[|%a|]"? (This is the failing path of the test, so there would be no need to update the reference file).

dra27 replied:
Sure - I just (pedantically) mapped it to the characters you'd used!

gasche replied:
(I also find `[|%a|]` more readable.)

stedolan commented on Makefile:
```diff
@@ -304,50 +304,51 @@ endif
+	    $(call SUBST,CC_PROFILE) \
+	    $(call SUBST,EXT_ASM) \
+	    $(call SUBST,EXT_DLL) \
 	    -e 's|%%EXT_EXE%%|$(EXE)|' \
```
Any reason why EXT_EXE doesn't get the same treatment?

dra27 replied:
It's the only one where the substitution variable name doesn't match the Makefile variable name... I guess `utils/config.mlp` could be changed...

stedolan commented:
license header on `tools/check-symbol-names` looks fine, thanks!

shindere commented:
David Allsopp (2017/08/14 11:22 +0000):
> dra27 commented on this pull request.
> 
> 
> 
> > -	    -e 's|%%CCOMPTYPE%%|$(CCOMPTYPE)|' \
> -	    -e 's|%%CC_PROFILE%%|$(CC_PROFILE)|' \
> -	    -e 's|%%EXT_ASM%%|$(EXT_ASM)|' \
> -	    -e 's|%%EXT_DLL%%|$(EXT_DLL)|' \
> +	sed $(call SUBST,AFL_INSTRUMENT) \
> +	    $(call SUBST,ARCH) \
> +	    $(call SUBST,ARCMD) \
> +	    $(call SUBST,ASM) \
> +	    $(call SUBST,ASM_CFI_SUPPORTED) \
> +	    $(call SUBST,BYTECCLIBS) \
> +	    $(call SUBST,BYTERUN) \
> +	    $(call SUBST,CC) \
> +	    $(call SUBST,CCOMPTYPE) \
> +	    $(call SUBST,CC_PROFILE) \
> +	    $(call SUBST,EXT_ASM) \
> +	    $(call SUBST,EXT_DLL) \
>  	    -e 's|%%EXT_EXE%%|$(EXE)|' \

Why this change, actually? What does it bring compared to how things
were done before?

Autoconf will make all this part go away anyway so I am not sure this
really needs to be touched?

dra27 commented:
@shindere - for this pull request, it means that `Makefile` now passes `tools/check-typo` as the `WITH_SPACETIME_CALL_COUNTS` options was tricky to split.

Semantically, it does also decrease the chance of generating invalid strings because it means that `\` is always correctly escaped for each variable. Previously this was done manually for `FLEXLINK_FLAGS`, `MKDLL`, `MKEXE` and `MKMAINDLL` only but in theory one could have paths in any of the compiler flags variables too.

shindere commented:
David Allsopp (2017/08/16 13:04 +0000):
> @shindere - for this pull request, it means that `Makefile` now passes
> `tools/check-typo` as the `WITH_SPACETIME_CALL_COUNTS` options was
> tricky to split.

Okay

> Semantically, it does also decrease the chance of generating invalid
> strings because it means that `\` is always correctly escaped for each
> variable. Previously this was done manually for `FLEXLINK_FLAGS`,
> `MKDLL`, `MKEXE` and `MKMAINDLL` only but in theory one could have
> paths in any of the compiler flags variables too.

Fine with me then. Thanks!

edwintorok reviewed (approved):
Changes to lintapidiff look fine

dra27 committed:
Ignore very long URLs in tools/check-typo

This is only to allow for a few specific instances, so for the time
being implemented for long-long and not very-long-line.

dra27 commented:
@gasche - thanks! I found a way to alter the ASCIIdoc output to reduce the line length with the URL. There's no equivalent in Markdown, and it seems a shame to enable `long-line` just for that one line. I've pushed a ~~little~~big further hack to check-typo which permits the ignoring of a long URL(ish) at the end of a line.

gasche commented on testsuite/tests/lib-digest/md5.ml:
```diff
@@ -120,10 +120,10 @@ let string_to_data s =
+      Int32.(logor (shift_left (of_int (Bytes.get s (j+3) |> Char.code)) 24)
+        (logor (shift_left (of_int (Bytes.get s (j+2) |> Char.code)) 16)
+          (logor (shift_left (of_int (Bytes.get s (j+1) |> Char.code)) 8)
+                 (of_int (Bytes.get s j |> Char.code)))))
```
Is there a way to format this code to make it readable?

```
let byte n = Bytes.get s (j+n) |> Char.code |> Int32.of_int in
let open Int32 in
byte 0
|> logor (shift_left (byte 1) 8)
|> logor (shift_left (byte 2) 16)
|> logor (shift_left (byte 3) 24)
```

dra27 replied:
Readable test cases?! Whatever next!

Sorry, I'd missed this one previously, but I've changed that - far better to improve the readability of the code, than just reduce the line lengths below the maximum.

gasche commented on typing/HACKING.adoc:
```diff
@@ -3,7 +3,7 @@ will need a good understanding of the OCaml type system and type
 typechecker:
 
-http://caml.inria.fr/pub/docs/u3-ocaml/index.html[Using, Understanding, and Unraveling the OCaml Language by Didier Rémy] ::
+http://caml.inria.fr/pub/docs/u3-ocaml/index.html[Using, Understanding, and Unraveling the OCaml Language by Didier R&eacute;my] ::
```
Why is this necessary, is this file not valid UTF8?

damiendoligez replied:
Funny fact: I've known Didier for more than 25 years, and I learned two weeks ago that his name officially doesn't have an accent...

gasche reviewed (approved):
Approved with minor comments.

damiendoligez commented:
> I've altered tools/check-typo to permit UTF-8 characters in copyright and authors lines in the headers (allows several existing headers to retain accented characters and also means no author in future should feel compelled to use incorrect/alternate spellings of names just to satisfy a script)

This is slightly dangerous: I've once seen a file with UTF-8 in the comments and Latin-1 in the code. Emacs got confused by this and introduced mojibake in the code (which made it buggy).

But if we make sure there's no non-ascii character anywhere else than in headers, I guess it's OK.

dra27 committed:
Warn if a file contains UTF-8 and Latin-1

Add a new warning non-ascii-utf8 displayed only if the non-ascii
attribute is specified and UTF-8 characters were ignored in the
copyright or authors lines in the header.

dra27 commented:
@damiendoligez - I've pushed an extra commit (https://github.com/ocaml/ocaml/pull/1287/commits/316319df07d5a70ce0794b8e92044c5a2daa55e5) which adds a `non-ascii-utf8` warning which activates only if the `non-ascii` attribute has been added for the file and a both a non-ASCII character is found *and* a UTF-8 sequence is skipped in the headers. I eventually managed to persuade my editor to concoct a test file for it 🙂

dra27 commented:
This is now rebased. I have hopefully applied enough care in the root Makefile (the change for updating utils/config.ml) to reflect recent additions, but this must therefore be allowed to run through CI.

All previous review points had been dealt with so as long as Damien is happy with the new UTF-8/non-ASCII check, I think this it then good to go, and I'll get #1288 updated. (I'm liking having tools/check-typo always check my commits, but I'm liking the fact it reveals other people's violations less!!)

damiendoligez commented on Makefile:
```diff
@@ -297,63 +297,71 @@ ifeq "$(FLEXDLL_SUBMODULE_PRESENT)" ""
   BOOT_FLEXLINK_CMD = FLEXLINK_CMD="../boot/ocamlrun ../flexdll/flexlink.exe"
   CAMLOPT := OCAML_FLEXLINK="boot/ocamlrun flexdll/flexlink.exe" $(CAMLOPT)
-  FLEXDLL_DIR=$(if $(wildcard flexdll/flexdll_*.$(O)),"+flexdll")
+  FLEXDLL_DIR=$(if $(wildcard flexdll/flexdll_*.$(O)),+flexdll)
```
Travis complains about this change. The value of FLEXDLL_DIR must be a string in OCaml syntax, so the quotes are needed.

dra27 replied:
It's not quite this - it's a mistake in one of the macros further down. I've just pushed a fix. The issue is that FLEXDLL_DIR is non-empty then it must be a complete OCaml string (with the double quotes) but if it's empty then it should be empty - as the result goes into a `string list` into sys/config.mlp. The new generic SUBST macro escapes double-quotes which was what was causing the problem for this specific variable.

shindere commented:
@dra27:

"Permit lines between 80-132 in .travis-ci.sh": why is that needed? Why
isn't it possible to have this file respect the 80 characters line
limit?

"Tweak .gitattributes for fakeclock.c in testsuite": same question
here, why this?

I thought we decided not to have copyright headers in test programs, but
perhaps I am mistaken here?

"Permit UTF-8 characters in copyright and authors": are we really sure
we want to go in this direction?

"Ignore very long URLs in tools/check-typo": typo in the commit message:
This is only to allow **a allow a** very long URL

Am I correct that you will get rid of the URL shortener, finally?

shindere commented:
David Allsopp (2017/10/11 10:39 +0000):
> This is now rebased. I have hopefully applied enough care in the root
> Makefile (the change for updating utils/config.ml) to reflect recent
> additions, but this must therefore be allowed to run through CI.

Will you use precheck?

dra27 commented:
@shindere: yes, I will run this through precheck (once it also passes Travis - careless of me not to have checked Linux in the first place).

shindere commented:
David Allsopp (2017/10/11 15:03 +0000):
> @shindere: yes, I will run this through precheck (once it also passes
> Travis - careless of me not to have checked Linux in the first place).

Great, thanks. Please let me know how it feels with precheck, I had
re-build it recently.

dra27 commented:
@shindere:

> "Permit lines between 80-132 in .travis-ci.sh": why is that needed? Why
isn't it possible to have this file respect the 80 characters line
limit?

The particular line was the API query because the URL is very long. Specifically for the CI, it seemed easier just to allow this one to have a few overly long lines because it should change very infrequently. I can revisit this if you prefer.

> "Tweak .gitattributes for fakeclock.c in testsuite": same question
here, why this?
>
>I thought we decided not to have copyright headers in test programs, but
perhaps I am mistaken here?

We had, but fakeclock.c is a massive piece of unscrupulous C on which I wish to be identified (both for the "oh my God, who did this" reason and also for copyright!).

> "Permit UTF-8 characters in copyright and authors": are we really sure
we want to go in this direction?

It's easy for me to say no one should have accents in their names! That said, I think we should permit it, yes, even though the support here is not perfect (part of the "TODO Ensure the file is valid UTF-8" part would also include "and ensure that the user has not used any combining characters", i.e. that it's normalised, but I think that then strays into over-engineering... and I'm certainly not planning on implementing Unicode normalisation in awk!).

> "Ignore very long URLs in tools/check-typo": typo in the commit message:
This is only to allow **a allow a** very long URL

Fixed :blush:

> Am I correct that you will get rid of the URL shortener, finally?

I should have re-worded this when I did the fixup (I have now) - I've used as many markdown/ASCIIdoc tricks as possible to shorten those lines.

shindere commented:
David Allsopp (2017/10/11 15:16 +0000):
> > "Permit lines between 80-132 in .travis-ci.sh": why is that needed? Why
> isn't it possible to have this file respect the 80 characters line
> limit?
> 
> The particular line was the API query because the URL is very long.
> Specifically for the CI, it seemed easier just to allow this one to
> have a few overly long lines because it should change very
> infrequently. I can revisit this if you prefer.

Perhaps it would be possible in check-typo to allow lines above 80
characters only for URLs?

I'm sorroy, I confused this file and appveyor_build.sh which also seems
to have several lines that go beyond 80 characters and could benefit
from a re-formatting.

> > "Tweak .gitattributes for fakeclock.c in testsuite": same question
> here, why this?
> >
> >I thought we decided not to have copyright headers in test programs, but
> perhaps I am mistaken here?
> 
> We had, but fakeclock.c is a massive piece of unscrupulous C on which
> I wish to be identified (both for the "oh my God, who did this" reason
> and also for copyright!).

I think file authors can already be identified through git blame and am
not very tnthousiastic about exceptions but if our fellow developers are
okay, fine with me.

> > "Permit UTF-8 characters in copyright and authors": are we really sure
> we want to go in this direction?
> 
> It's easy for me to say no one should have accents in their names!
> That said, I think we should permit it, yes, even though the support
> here is not perfect (part of the "TODO Ensure the file is valid UTF-8"
> part would also include "and ensure that the user has not used any
> combining characters", i.e. that it's normalised, but I think that
> then strays into over-engineering... and I'm certainly not planning on
> implementing Unicode normalisation in awk!).

Well I wouldn't have changed anything, although I am concernend. I think
I'd prefer to keep things as simple as possible.

> > "Ignore very long URLs in tools/check-typo": typo in the commit message:
> This is only to allow **a allow a** very long URL
> 
> Fixed :blush:
> 
> > Am I correct that you will get rid of the URL shortener, finally?
> 
> I should have re-worded this when I did the fixup (I have now) - I've
> used as many markdown/ASCIIdoc tricks as possible to shorten those
> lines.

OK.

dra27 commented:
This is presently going through precheck: https://ci.inria.fr/ocaml/job/precheck/17/

@shindere - I'll come back on your comments later (hopefully with some changes), but reformatting CI scripts is a low-risk - the main thing is to get precheck to verify the Makefile changes.

dra27 committed:
Reformat .travis-ci.sh

dra27 commented:
@shindere - you're absolutely right about `.travis-ci.sh` ... a case of iterative development - I added the stuff for skipping a URL after I'd made the change to .gitattributes!

I've removed the commit from here, and my working version of #1288 refactors the script entirely.

I also spotted an unnecessary variable in the changes for generating utils/config.ml so it's just running through CI again.

damiendoligez commented:
It's OK to have a few files with copyright headers in testsuite. What I don't like is having a whole lot of files with a 20-line header followed by only a few lines of code.

dra27 commented:
This is looking OK in precheck: https://ci.inria.fr/ocaml/job/precheck/18/

dra27 commented:
Not sure why ocaml-ppc-64 flambda and all 4 macos slaves are pending even overnight on precheck, but this does all seem to be fine.

Is this ready to merge (I would suggest rebase, rather than squash)? I'm happy to do the cherry-pick onto 4.06, as I expect it'll be fiddly, so worth running through precheck itself beforehand.

One last rebase - the changes were @gasche's suggestion for the MD5 test (so just a quick CI pass required here), a fix of a typo in a commit message and an expansion of the comments in check-typo about UTF-8.

xavierleroy commented:
> Not sure why ocaml-ppc-64 flambda and all 4 macos slaves are pending 

The macos slaves are offline for reasons unknown to me.  ocaml-ppc-{32,64} flambda is not run because too time-consuming for the aging G5 that runs them.  Jenkins displays them in different shades of gray and blue for unknown reasons, but they don't block the build, unlike the macos slaves...

dra27 commented:
@xavierleroy - thanks! It would be reassuring if someone on a Mac could just run `./configure && make world` just to be sure that the changes to utils/config.ml haven't subtly broken on macos (I'm remembering a previous change which looked fine and then turned out to be very subtly, but noticeably, broken on OS X 10.8, or something...)

damiendoligez commented:
I'm building both trunk and 4.06 quite regularly on my mac, so build system goofs are covered.

The Mac slaves are offline (for what I hoped would be an afternoon but will turn out to be a few days) in order to install a new slave with the latest version of macOS, High Sierra.
I'll restart one of them now to minimize the inconvenience.

dra27 commented:
I'm somewhat baffled by the flambda failures in https://ci.inria.fr/ocaml/job/precheck/20/

@shindere / @xavierleroy / @damiendoligez - is it possible to restart specific parts of a job, or the whole thing have to re-run through precheck?

shindere commented:
David Allsopp (2017/10/12 20:30 +0000):
> I'm somewhat baffled by the flambda failures in
> https://ci.inria.fr/ocaml/job/precheck/20/

Investigating this currently, thanks for the report!

> @shindere / @xavierleroy / @damiendoligez - is it possible to restart
> specific parts of a job, or the whole thing have to re-run through
> precheck?

Currently it isn't but there is a Jenkins plugin that makes it possible.
Will install it ASAP.

xavierleroy commented:
I've seen those failures before.  It looks like a race condition that garbles the output of the tests and therefore confuses the script that tallies the failed tests at the end.  Example of garbled output:
```
 ... testing 'nucleic.ml': ocamlcRunning tests from 'tests/array-functions' ...
 ... testing 'test.ml' with 1 (native) => passed
```
Then we get the conclusion:
```
Summary:
  250 tests passed
   13 tests skipped
    0 tests failed
    1 unexpected errors
  264 tests considered

List of unexpected errors:
    tests/array-functions/nucleic.ml
```

Feel free to re-run the whole test, it "only" takes 1 hour.

shindere commented:
I just installed Jenkins' Matrix Reloaded plugin
(https://wiki.jenkins.io/display/JENKINS/Matrix+Reloaded+Plugin).

It won't help for the builds that already occurred like precheck #20,
but for future builds it should provide a "Matrix reloaded" link that
lets one restart failed builds.

dra27 commented:
@xavierleroy - ah, I see - I assumed it was a console display issue, not that the log is actually corrupted. Weird, but I guess we're gradually moving towards a better system for that anyway. I've scheduled [precheck#21](https://ci.inria.fr/ocaml/job/precheck/21/)!

@shindere - cool, thanks for that!

dra27 commented:
Assuming this does pass precheck, is everyone happy for this to be merged?

@shindere - I realise hadn't answered your point about the accents in the headers. In this case, I was altering check-typo to handle accents in files which are *already* checked in with accents (e.g. asmcomp/linscan.ml), rather than editing anyone else's name to add "missing" accents. All the changes here lead towards the bigger, more mechanical #1288 which updates the entire repository to pass check-typo, and so makes using the githook in #1148 a practical thing to recommend.

shindere commented:
David Allsopp (2017/10/13 01:14 -0700):
> @shindere - I realise hadn't answered your point about the accents in
> the headers. In this case, I was altering check-typo to handle accents
> in files which are *already* checked in with accents (e.g.
> asmcomp/linscan.ml), rather than editing anyone else's name to add
> "missing" accents.

OK. Didn't we want to *remove* the accents, actually?

mshinwell commented:
@dra27 Can we try to do the final push to get this merged?  It looks as if it's very close.

damiendoligez commented:
> OK. Didn't we want to *remove* the accents, actually?

I've given up on removing accents from the names of contributors and institutions.

damiendoligez reviewed (approved):
I have reviewed the changes and they look good.

dra27 commented:
OK - I'll rebase this now

dra27 commented:
Done - the change in Makefile was to rename `WITH_SPACETIME_CALL_COUNTS` to `ENABLE_CALL_COUNTS` from 8f6c6aca926bd7b7f221e2032677040219c5d2d8

mshinwell commented:
@dra27 Please write a Changes entry and I think you can merge this; @damiendoligez approved yesterday.

shindere commented:
Damien Doligez (2017/10/20 02:41 -0700):
> > OK. Didn't we want to *remove* the accents, actually?
> 
> I've given up on removing accents from the names of contributors and
> institutions.

:(

mshinwell commented:
@dra27 Is there a reason this hasn't been merged to 4.06?  You just need to write a Changes entry, or so I thought.

dra27 commented:
Travis Changes test appears to be overloaded. Merged and cherry-picked to 4.06 as 4e03f871212aaba3eb2a216bfbb36838f65b7dbf
